include/openenclave/bits/seal.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+/**
+ * @file seal.h
+ *
+ * This file defines constants and structures for sealing APIs.
+ *
+ * Only TEE agnostic definitions should go to this file. SGX specific
+ * definitions should go to sgx/seal.h
+ */
+
+#ifndef _OE_BITS_SEAL_H
+#define _OE_BITS_SEAL_H
+
+#include <openenclave/bits/evidence.h>
+#include <openenclave/bits/types.h>
+
+OE_EXTERNC_BEGIN
+
+typedef struct _oe_seal_setting_header
+{
+    int name;
+    size_t size;
+} oe_seal_setting_header_t;
+
+/**
+ * TEE agnostic seal settings supported by OE SDK
+ */
+typedef enum _oe_seal_setting_name
+{
+    OE_SEAL_SETTING_POLICY,
+    OE_SEAL_SETTING_ADDITIONAL_CONTEXT,
+    OE_SEAL_SETTING_TEE_AGNOSTIC_MAX
+} oe_seal_setting_name_t;
+
+typedef struct _oe_seal_setting_policy
+{
+    const oe_seal_setting_header_t header;
+    oe_seal_policy_t policy;
+} oe_seal_setting_policy_t;
+
+typedef struct _oe_seal_setting_additional_context
+{
+    const oe_seal_setting_header_t header;
+    const void* buffer;
+    size_t size;
+} oe_seal_setting_additional_context_t;
+
+#if __x86_64__ || _M_X64
+#include <openenclave/bits/sgx/seal.h>
+#endif
+
+/**
+ * Seal data to an enclave using AEAD (Authenticated Encryption with Additional
+ * Data).
+ *
+ * @param[in] plugin_id Optional UUID of the plugin to use. If \c NULL, the
+ * default plugin will be used.
+ * @param[in] settings Optional array of seal settings to be used.
+ * @param[in] settings_count The number of settings specified by \p settings.
+ * Must be \c 0 if \p settings is \c NULL.
+ * @param[in] plaintext Optional buffer to be encrypted under the seal key.
+ * @param[in] plaintext_size Size of \p plaintext, must be \c 0 if \p plaintext
+ * is \c NULL.
+ * @param[out] blob On success, receive the pointer to a buffer containing both
+ * \p additional_data and encrypted \p plaintext, along with necessary
+ * information for unsealing. Freed by \c oe_free().
+ * @param[out] blob_size On success, receive the size of \p blob.
+ *
+ * @retval OE_OK \p plaintext was sealed to the enclave successfully.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
+ * @retval OE_CRYPTO_ERROR Error occurred during encryption.
+ */
+oe_result_t oe_seal(
+    const oe_uuid_t* plugin_id,
+    const oe_seal_setting_header_t* settings[],
+    size_t settings_count,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    uint8_t** blob,
```

Comment:
Here the question is where to store AAD. 2 options:

1. AAD stored as part of `blob`.
2. AAD included in `blob`'s MAC only. The caller must store AAD separately and provide it to `oe_unseal()` for MAC verification.

We ruled out (1) because the caller doesn't understand `blob`'s format so cannot extract ADD from it (hence cannot make use of AAD without unsealing). 

If anyone thinks (2) is valuable then I can add them in. I don't have a strong opinion here.

Reply:
FWIW, I'm fine with the design where additional_data is not automatically incorporated into the resulting blob.